### PR TITLE
GRPO loss fix

### PR DESCRIPTION
**Pull request by vwxyzjn**
CC @gauravpandeyamu @hamishivi for a second look. 

I think the `pg_loss = pg_loss + (args.beta * kl).mean()` is incorrect because the mean operation happens after the scores are added with the corresponding KL

image

The correct impl should be 

```
pg_loss = masked_mean(pg_loss_max + (args.beta * kl), ~padding_mask[micro_batch_inds])
```

Let me know what you think.

### CODE REVIEW

**Comment by gauravpandeyamu** [comment]
Yes, masked_mean is indeed correct to use here.

Also check my comment at https://github.com/allenai/open-instruct/issues/534#issuecomment-2624863523

The current kl1 estimator will completely ignore the ref_logprobs during trainining. kl2 and kl3 are fine.

File: open_instruct/grpo_vllm_thread_ray_gtrl.py [hunk 1]
```diff
@@ -487,6 +487,34 @@ def remove_padding(sequences, pad_token_id):
     return [[inneritem for inneritem in item if inneritem != pad_token_id] for item in sequences]
 
 
+class MetricsTracker:
```
Comment (hamishivi) [inline open_instruct/grpo_vllm_thread_ray_gtrl.py:490]: nice!

File: open_instruct/grpo_vllm_thread_ray_gtrl.py [hunk 2]
```diff
@@ -1170,25 +1206,32 @@ def vllm_generate(
...
-                        # recalculate kl: difference from PPO because we want the KL loss
-                        # to backpropagate through the model
-                        # TODO: we potentially need to clip this kl loss in case some logits
-                        # exploded, but more on it later.
-                        kl1 = new_logprobs - ref_logprobs[micro_batch_inds]
-                        kl2 = (kl1) ** 2 / 2
-                        kl3 = (-kl1).exp() - 1 + kl1
+                        # Here we recalculate kl: we want the KL loss to backpropagate through the model
+                        # We also clamp the KL loss to avoid numerical instability
+                        # https://chatgpt.com/c/679c4aa5-a558-8011-a1de-142381105e56
```
Comment (hamishivi) [inline open_instruct/grpo_vllm_thread_ray_gtrl.py]: I can't load this conversation. Maybe needs to be made public or rehost somewhere?

**Review by hamishivi — APPROVED** [review]
Looks good to me. Nice on the metrics improvement!
